### ISSUE TITLE
How to track down 'Stores must be a multiple of 8 bits'

### ISSUE THREAD

**Issue by janjongboom** [kripken/emscripten#7595]
@kripken I'm cross-compiling a huge code base, and I'm running into an error in the LLVM JSBackend. In the `splitStore` function a `Stores must be a multiple of 8 bits` error is thrown. The value of ` DL->getTypeSizeInBits(Inst->getValueOperand()->getType())` is `2`, so that makes sense, but I'm wondering how best to find out where this type is declared. I'm new to LLVM and diving into the `Type` class doesn't yield me any declaration info, so I'm wondering if this information is already lost at this stage.

Alternatively, a pointer on how to attach a debugger to LLVM would be useful too.

**Comment by kripken** [comment]
Probably simplest to run `llvm-dis` on the input bitcode, and grep to find where an `i2` appears in the text IR.

However, at this point it may be a good idea to just test with the LLVM wasm backend instead of the JSBackend. We hope to turn it on by default soon, and it can handle `i2` and other illegal values.

For attaching a debugger, run with `EMCC_DEBUG=1` to see the LLVM commands being issued. You can then run them in the debugger directly.

**Comment by janjongboom** [comment]
Just a follow up, as it's one of the weirdest bugs / optimizations that I've debugged in a while. I only saw this error when doing an optimized build (-O2), and it pointed to a function which looked very straightforward. Just calling some functions and then a straightforward `if` clause ([this one](https://github.com/ARMmbed/mbedtls/blob/fabc6001ff1cf2132550e34248b5f16f42b7e86d/library/x509_crt.c#L1188)). 

This is the LLVM bitcode that is emitted:

```
  %trunc = trunc i32 %.0 to i2
  switch i2 %trunc, label %47 [
    i2 0, label %.backedge
    i2 -2, label %.backedge
    i2 -1, label %.thread70
  ]
```

The potential values that are checked in the `if` clause are `-0x1080`, `-0x1480` and `0x0` and that's where some compiler optimizations seem to kick in. I think that it does a truncate and then only checks the bits that can differ. In binary they look like:

```
-0x1080
11101111 10000000

-0x1480
11101011 10000000

0x0
00000000 00000000
```

I believe the compiler sees the clauses, decides that it distinguish the three values by looking at just the three (or two?) bits that differ between them on the first byte (well, third, as this is int32), and issues a truncate from i32 to i2, and then does a switch on the truncated value. Inverting the if clause solved the issue.
